### PR TITLE
[ML-9890] Add tests for all primitive types

### PR DESCRIPTION
**Pull request by liangz1**
df: DataFrame | rows=df.collect() | Tensor (in Dataset) | Evaluated tensor dtype | Evaluated tensor element type
-- | -- | -- | -- | --
BooleanType | bool | bool | np.bool_ | bool_
ByteType | int | int8 | np.int8 | int8
ShortType | int | int16 | np.int16 | int16
IntegerType | int | int32 | np.int32 | int32
LongType | int | int64 | np.int64 | int64
FloatType | float | float32 | np.float32 | float32
DoubleType | float | float64 | np.float64 | float64
StringType | str | string | np.object_ | bytes
BinaryType | bytearray | string | np.object_ | bytes

### What changes are proposed in this PR?
Add tests for primitive types that petastorm supports. Added ByteType, StringType and BinaryType.
### Why these changes are needed?
These tests make it clear which column type is mapped to what types in the tensor.

### CODE REVIEW

File: petastorm/tests/test_spark_dataset_converter.py [hunk 1]
```diff
@@ -24,11 +24,15 @@ def test_primitive(self):
...
-            StructField("long_col", LongType(), False)
+            StructField("long_col", LongType(), False),
+            StructField("str_col", StringType(), False),
+            StructField("bin_col", BinaryType(), False),
+            StructField("byte_col", ByteType(), False),
         ])
         df = self.spark.createDataFrame([
-            (True, 0.12, 432.1, 5, 5, 0),
-            (False, 123.45, 0.987, 9, 908, 765)], schema=schema).coalesce(1)
+            (True, 0.12, 432.1, 5, 5, 0, "hello", bytearray(b"spark"), -128),
```
Comment (WeichenXu123) [inline petastorm/tests/test_spark_dataset_converter.py]: You can use `b"spark"` instead of `bytearray(b"spark")`.

Comment (WeichenXu123) [reply petastorm/tests/test_spark_dataset_converter.py]: We'd better test some special characters(arbitrary bytes) like 
`b"spark\x01\x02spark"`

File: petastorm/tests/test_spark_dataset_converter.py [hunk 2]
```diff
@@ -42,16 +46,24 @@ def test_primitive(self):
                 # Now we only have one batch.
             for i in range(converter.dataset_size):
                 for col in df.schema.names:
-                    self.assertEquals(getattr(ts, col)[i], expected_df[i][col])
+                    actual_ele = getattr(ts, col)[i]
+                    expected_ele = expected_df[i][col]
+                    if type(actual_ele) == bytes:
+                        actual_ele = actual_ele.decode()
```
Comment (WeichenXu123) [inline petastorm/tests/test_spark_dataset_converter.py:54]: We should directly compare "bytes" type in python. The decode is used when there's encoding like utf8. Here the "Binary" type in spark support arbitrary binary content, such as raw file content bytes.

compare like:
`bytes1 == bytes2` or `bytearray1 == bytes2` will work

File: petastorm/tests/test_spark_dataset_converter.py [hunk 3]
```diff
@@ -42,16 +46,24 @@ def test_primitive(self):
...
+                    if type(expected_ele) == bytearray:
+                        expected_ele = expected_ele.decode()
```
Comment (WeichenXu123) [inline petastorm/tests/test_spark_dataset_converter.py]: I think the df.collect will always return "bytes" types, so we can remove this branch code.

File: petastorm/tests/test_spark_dataset_converter.py [hunk 4]
```diff
@@ -42,16 +46,24 @@ def test_primitive(self):
...
-            self.assertEquals(len(converter), len(expected_df))
+            self.assertEqual(len(converter), len(expected_df))
 
-        self.assertEquals(ts.bool_col.dtype.type, np.bool_, "Boolean type column is not inferred correctly.")
-        self.assertEquals(ts.float_col.dtype.type, np.float32, "Float type column is not inferred correctly.")
-        self.assertEquals(ts.double_col.dtype.type, np.float64, "Double type column is not inferred correctly.")
-        self.assertEquals(ts.short_col.dtype.type, np.int16, "Short type column is not inferred correctly.")
-        self.assertEquals(ts.int_col.dtype.type, np.int32, "Integer type column is not inferred correctly.")
-        self.assertEquals(ts.long_col.dtype.type, np.int64, "Long type column is not inferred correctly.")
+        self.assertEqual(ts.bool_col.dtype.type, np.bool_, "Boolean type column is not inferred correctly.")
```
Comment (WeichenXu123) [inline petastorm/tests/test_spark_dataset_converter.py]: What's the method `bool_col`, `float_col` mean ? Is there some doc ?

File: petastorm/tests/test_spark_dataset_converter.py (hunk 3 above)
Comment (WeichenXu123) [reply petastorm/tests/test_spark_dataset_converter.py]: oh, I am wrong, spark always generate results as "bytearray" type. So remove the "bytes" branch.

File: petastorm/tests/test_spark_dataset_converter.py (hunk 1 above)
Comment (liangz1) [reply petastorm/tests/test_spark_dataset_converter.py]: Using b'spark' doesn't work in CI tests... but it works locally.

File: petastorm/tests/test_spark_dataset_converter.py (hunk 2 above)
Comment (liangz1) [reply petastorm/tests/test_spark_dataset_converter.py:54]: For `StringType` column, the `actual_ele` will be `b'hello'` (bytes: 5) while `expected_ele` is `'hello'` (str).

Comment (liangz1) [reply petastorm/tests/test_spark_dataset_converter.py:54]: I summarized the mapping in the table in the PR description. The `Evaluated tensor element type` is the type of `actual ele` and the `rows=df.collect()` is the type of `expected_ele`.

File: petastorm/tests/test_spark_dataset_converter.py (hunk 3 above)
Comment (liangz1) [reply petastorm/tests/test_spark_dataset_converter.py]: The bytes branch is for `actual_ele`, not `expected_ele`.

File: petastorm/tests/test_spark_dataset_converter.py (hunk 2 above)
Comment (liangz1) [reply petastorm/tests/test_spark_dataset_converter.py:54]: For BinaryType, I cast `actual_ele` from `bytes` to `bytearray`. For StringType, I decode `actual_ele` from `bytes` to `str` using default decode=`ascii`.

Comment (WeichenXu123) [reply petastorm/tests/test_spark_dataset_converter.py:54]: oh, looks good. But if it is string type, the `getattr(ts, col)[i]` should already been a str object, I guess tensorflow already decode it ?

Comment (WeichenXu123) [reply petastorm/tests/test_spark_dataset_converter.py:54]: Oh, you're right, tf.string type element always is b'xxx' type so need decode.

**Review by WeichenXu123 — APPROVED** [review]

**Pull request merged**